### DISCUSSION TITLE
[android] image button.on_press() possible?

### DISCUSSION THREAD

**Discussion by HKLM** [beeware/toga#1343] [Q&A]
Is there any way to make a button, that instead of text, display a icon or image? (That will work on android)
Im looking for a way to get a "on_press" event from a icon or image

**Comment by freakboy3742** [comment]
This is something that has come up a few times - see [this issue](https://github.com/beeware/toga/issues/774) for at least one of those discussions.

The short answer to your question is that this isn't currently possible. It's been proposed in PR #1213, but I'm still not sold on the use case.

The underlying question is *why* do you want this? If what you want is a button with an icon, then we don't need an image that supports on_press - we need to support icons on buttons. This is something I can definitely support - and at some level, should already be possible by using unicode in the button label.

A button is very different to an image. An image is a picture, of arbitrary size. There might be a valid use case for making the image clickable - but being able to fake a button isn't that use case.

**Reply by HKLM** [reply]
> This is something that has come up a few times - see [this issue](https://github.com/beeware/toga/issues/774) for at least one of those discussions.
> 
> The short answer to your question is that this isn't currently possible. It's been proposed in PR #1213, but I'm still not sold on the use case.
> 
> The underlying question is _why_ do you want this? If what you want is a button with an icon, then we don't need an image that supports on_press - we need to support icons on buttons. This is something I can definitely support - and at some level, should already be possible by using unicode in the button label.
> 
> A button is very different to an image. An image is a picture, of arbitrary size. There might be a valid use case for making the image clickable - but being able to fake a button isn't that use case.

I would like to have a button with a icon displayed on it. I think it would make my app visually more appealing. Right now it reminds me of a webpage from 1990's :) . For example, to display a gear icon that opens up a settings window. vs a button that says "settings"

I have a working prototype android app that can tell me if my doors are locked/unlocked, windows are up or down, I can remotely lock and unlock my truck. The app Ford has out does not tell me the status of my windows or doors or even if the alarm is active or not. Even tho I am using Ford's owns RESTful API to do this, my app can do it.  I'm not a professional programmer. In fact this is my 1st time using python, I've more experience with C#. In fact I was originally trying to get this to work using Xamarin Forms. But then I came across beeware and so far it has fulfilled almost everything I want to do.  I'm now at the point where I am trying to polish it up and make it nice and clean visually.  For example: I have 2 buttons "Update" and "Lock". Not very elegant and boring.  I have these two icons I would rather have displayed

"Update" would be 

![sharp_sync_black_18dp](https://user-images.githubusercontent.com/10490692/137226935-048a60d7-0a85-4d5b-8942-2a8f65b63c48.png)

And "Lock" would be
![baseline_lock_outline_black_36](https://user-images.githubusercontent.com/10490692/137226969-ddc8a1ae-730d-4618-b852-200550dbf1fe.png)


And thus tapping on either one do the on_press() and so on.  Currently there is no way to do this on android that I know of.  All the widgets that use icons, with the exception of imageview, don't work on android.  There is no command widget for android.  I'm not asking for a major reworking of toga.  I just want to know any way that this can be implemented in my app? 

[Android has a native ImageButton widget. ](https://developer.android.com/reference/android/widget/ImageButton)
I think you guys have a great thing here. It works, I like it and think it has a lot of potential. But I don't get why there is a reluctance to be able to have a button with a icon on it or a clickable image, or even a on_press event for a imageview widget or even `<a href=xxx><img></a>` or what ever.  I just want a icon that when clicked on, fires off a specified function. 

I don't like having to beg for something.
If this is not the direction you want toga to be going, then I'm going to look for other ways to do this.  

This is my app currently
![Screenshot_20211013-154157](https://user-images.githubusercontent.com/10490692/137222829-b8ffcae7-8ce8-4403-95a6-9900bb279a04.png)


This is Ford's official app that I can use to remotely lock my truck.
![Screenshot_20211013-154255](https://user-images.githubusercontent.com/10490692/137222830-5a29a1dc-fceb-4616-afca-83830370e7c3.png)
.

**Comment by freakboy3742** [comment]
My apologies - I think we've got some crossed wires here because of terminology.

In Toga's terminology, what you've described are *icons*, not images. Although it might not be obvious, there's a very important distinction here. Icons are:
 * small; 
 * square (or at least, NxN bitmaps), 
 * provided by the system or the app; and 
 * on many platforms, a single color (although this isn't true on every system)

 Images, on the other hand, are:
 * Arbitrarily sized (and can be *huge*, requiring scaling);
 * usually provided by the *user*;
 * have arbitrary colors.

Using the "official" app as an example - the start/lock/unlock pictures are *icons*. The picture of the truck is an *image*.

As a result, the APIs that we provide for icons vs images can be very different. 

What you've described - putting a "settings" or "refresh" *icon* on a button is (to me) an entirely reasonable use case, and one that we should definitely support. The app you're describing looks like a really solid example of where you'd use them (and, as an aside - the app itself seem amazing!). We sorta-kinda have support for this right now - in the form of using the icons that are available in Unicode [(such as the Media Control symbols)](https://en.wikipedia.org/wiki/Media_control_symbols) - however, those aren't a great solution; I'm only suggesting them as "something that will definitely work right now" because you can treat the icons as "text labels".

What I'm "resisting" is the idea of an *image* button. 

The reason for this is somewhat philosophical - I'm utterly sick of UI designers reinventing the wheel. Platforms have UI design guidelines for a reason, and consistency is something that *aids* user experience. Take the screenshot of the official app that you've provided. Based on the visual language, it's very difficult to know for certain what is and what isn't a button - because the design language is completely inconsistent. I'm guessing the big start icon is a button - they've used shadows to try and develop some sort of skeuomorphism that makes it look like a Big Button You Can Press. However, that skeuomorphism is inconsistent with the Android style guide... so that leads to things like the "24x7 assistance" pictogram. Is that a button? The design language isn't consistent with the other buttons on the screen, so how is the user meant to intuitively know it's a button. What about the "F-250 XLT F4 v" text? Is that button? Is it *all* a button, or just the down chevron? Who knows? The design language doesn't tell you. What about "Find a dealer"? I presume that's a button as well? What about the picture of the truck .... is that a button? Who knows? 

So - Ford has built a UI that has 3 (4?) buttons... and *they're all different* - **WHY**? Now, I'll grant that in this specific case, you might say it's not that big a deal. However, I've lost count of the number of times I've seen UIs where users *literally don't know that something is a button* because A Designer Has Designed The App; and while it may be "pretty" (for some definition of pretty...), how the app is actually meant to function is completely opaque. I've seen users completely surprised that a specific feature or functionality is available at all, because the fact that one particular "image" was actually an interactive element wasn't obvious.

What I'm trying to do with Toga's APIs is make it easy to do the right thing - which means being consistent with the platform style guide. That means buttons look like buttons, and behave like buttons. Depending on the platform, a button has an edge indicating how big it is, some contrast to show how big the pressable area is; simple text or an icon, and maybe a little shadow to give the visual impression that it can be pushed. An *image* is a picture. And while, yes, you can make the picture look like some sort of button - just because you've got a hammer, doesn't mean you need to treat everything like a nail.

When someone says "I want an image button", the question I ask is "do you *really* want an **Image** button, or do you want an **icon** on a button"? And, in my experience, the answer is almost always *icon*. I'm 100% on board with supporting *Icons* on buttons. We don't have support for them *yet*, but I'm not opposed to adding them. 

Internally, that might end up being implemented *on Android* using `android.widget.ImageButton` - but that's an implementation detail. Toga is an abstraction API - it doesn't expose the eccentricities of the underlying platform. iOS and macOS, for example, *don't* have an ImageButton - you can set an image on any button. When we expose a piece of functionality, we find an API abstraction that makes sense *in general*, rather than dogmatically following one specific platform API.

As for "begging" - well, I can tell you what features we have right now. I can also tell you that this is a volunteer-based Open Source project, so that addition won't happen unless someone volunteers to add it.

**Reply by HKLM** [reply]
> My apologies - I think we've got some crossed wires here because of terminology.
> 
> In Toga's terminology, what you've described are _icons_, not images. Although it might not be obvious, there's a very important distinction here. Icons are:
> 
> * small;
> * square (or at least, NxN bitmaps),
> * provided by the system or the app; and
> * on many platforms, a single color (although this isn't true on every system)
> 
> Images, on the other hand, are:
> 
> * Arbitrarily sized (and can be _huge_, requiring scaling);
> * usually provided by the _user_;
> * have arbitrary colors.

> What you've described - putting a "settings" or "refresh" _icon_ on a button is (to me) an entirely reasonable use case, and one that we should definitely support. The app you're describing looks like a really solid example of where you'd use them (and, as an aside - the app itself seem amazing!). We sorta-kinda have support for this right now - in the form of using the icons that are available in Unicode [(such as the Media Control symbols)](https://en.wikipedia.org/wiki/Media_control_symbols) - however, those aren't a great solution; I'm only suggesting them as "something that will definitely work right now" because you can treat the icons as "text labels".
> 
> What I'm "resisting" is the idea of an _image_ button.

Need I remind you, technically the icon widget in toga is currently not supported on android. So forgive me that I am using a .png file to display a image, "of a icon", on a button widget. 
![why_is_this_wrong](https://user-images.githubusercontent.com/10490692/140464620-7fdf42c8-3772-4ebf-a23a-cdc761bdf779.png)


> 
> The reason for this is somewhat philosophical - I'm utterly sick of UI designers reinventing the wheel. Platforms have UI design guidelines for a reason, and consistency is something that _aids_ user experience. 
> 
> What I'm trying to do with Toga's APIs is make it easy to do the right thing - which means being consistent with the platform style guide. That means buttons look like buttons, and behave like buttons. Depending on the platform, a button has an edge indicating how big it is, some contrast to show how big the pressable area is; simple text or an icon, and maybe a little shadow to give the visual impression that it can be pushed. An _image_ is a picture. And while, yes, you can make the picture look like some sort of button - just because you've got a hammer, doesn't mean you need to treat everything like a nail.
> 
Why is it ok for Google to style their own default clock app for android, but you don't want us to do the same with toga? WTF? 
Both have a small one color icon picture of a symbol, that when the user "presses it"/ "clicks it" it does something ([Google defines a Imagebutton](https://developer.android.com/reference/android/widget/ImageButton) as "Displays a button with an image (instead of text) that can be pressed or clicked by the user."
[Microsoft defines a ImageButton](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/imagebutton) as "The ImageButton displays an image and responds to a tap or click that directs an application to carry out a particular task.")
If it walks like a duck, and talks like a duck, its a duck,.   



> When someone says "I want an image button", the question I ask is "do you _really_ want an **Image** button, or do you want an **icon** on a button"? And, in my experience, the answer is almost always _icon_. I'm 100% on board with supporting _Icons_ on buttons. We don't have support for them _yet_, but I'm not opposed to adding them.
> 
> As for "begging" - well, I can tell you what features we have right now. I can also tell you that this is a volunteer-based Open Source project, so that addition won't happen unless someone volunteers to add it.

While you may have a dislike for the modern style of "buttons", I personally do not like the old Windows 3.1 style of a "button". Yes I've been using computers before Microsoft windows was even invented.  And I'm not the only one asking for this feature, there are a few others that have asked for the same feature. #774  #1257.  It bothers me to see that your holding your own project here back. You already have this button code #1213 done for you, its all ready to be added into the main branch. But you are not merging that PR....? Why?   Its like a father locking his daughter in her room for her entire adult life, because he does not like what she "might wear" if she was allowed out of the house.

Its time to let your daughter grow up

**Reply by johnzhou721** [reply]
Google's own "Image buttons" are in a tab bar — outside of that, in standalone "do this" buttons, these large icons aren't really shown a lot.

**Comment by clemsam** [comment]
There is a possibility _to make an **image(!)** "clickable"_ by using code from an active pull request. 
I used this PR's fork code _(**NOT** merged!)_ as described in this discussion https://github.com/beeware/toga/discussions/1372, BUT:

_**it does NOT behave like a button** (= change color on_press() ) !!!_

I myself do _NOT_  think an image on a (typically small) button to make much sense, btw:
~~_... how about allowing ".icon" files on buttons -- ?!?_~~ [Edit: _**IS** already "allowed" but not yet implemented._]

**Reply by HKLM** [reply]
> There is a possibility _to make an **image(!)** "clickable"_ by using code from an active pull request. I used this PR's fork code _(**NOT** merged!)_ as described in this discussion #1372, BUT:
> 
> _**it does NOT behave like a button** (= change color on_press() ) !!!_
> 
> I myself do _NOT_ think an image on a (typically small) button to make much sense, btw: _... how about allowing ".icon" files on buttons -- ?!?_

I did that a long time ago to make it work. What I dont understand here is why are the devs intentionally holding working features back, that people want? The feature of displaying a icon or image on a button. Im not the first person to ask about it. Like I said I found the working code for it and have completed my app, so I dont really care anymore. I just think it is a shame they are holding this back from others. I think they could have a really great platform here, if they would just allow for a icon, picture, or whatever to be clickable to call a method when clicked or pressed or touched, whatever, you get the idea.

**Reply by freakboy3742** [reply]
I don't know how much more clear I can be. I've now said *multiple* times that I'm 100% in favour of the idea of supporting *icons* on *buttons*. That doesn't mean I'm personally committing to adding that feature myself - but if the feature were to magically appear, I'd merge it.

However, as I've also said, that is *not* the same thing as adding an on_click handler to ImageView. I have also explained my rationale for that decision.

**Reply by clemsam** [reply]
@ freakboy3742 -- I am _very sorry_ that I didn't read your postings here carefully enough!
I know I sometimes get carried away with my own foolish "ideas" and am acting over-enthusiastic...
My apologies for that. 

I _VERY_ much appreciate this/_YOUR_ project !!!

_**IF** I was an IT expert/professional developer I would certainly  write the necessary code right away... but I'm not :-(  ._

**Reply by l-o-u-i-s-g** [reply]
@freakboy3742  I know I'm waking up an old thread. Maybe there´s newer development on this?  what do you think of the idea of simply adding a new widget.  Something like an EventBox  that would just extend the Box  widget  but have  an event listener  for input events.  If I understand this correctly  I believe thats how android does it  with the view class. https://developer.android.com/develop/ui/views/touch-and-input/input-events   that way  longpresses etc  could be added also.  So having an  on_press , on_long_press  etc  on an EventBox  would solve the case for every use case. The button could even be deprecated. 

Looks like pygtk already has an EventBox:  https://www.tutorialspoint.com/pygtk/pygtk_eventbox_class.htm

**Reply by freakboy3742** [reply]
@tiwwwi There's no new development, and I can't say my thinking on this matter has really changed that much. I'm 100% in favour of adding icons on buttons, but I'm not committing to implementing this myself at any point in the near future. 

Adding a new widget doesn't really solve anything (if anything, it creates a whole bunch of new problems, because now there's more widgets to support). 

The issue isn't *how* to implement the feature, it's *why* the feature is being requested. Ever single person who has raised a request for "ImageButton" has, in reality, been asking for one of two things:

* The ability to put an icon on a button
* An implementation of OptionContainer on their (mobile) platform.

I'm yet to hear a compelling argument for an *image* on a *button*.

**Reply by l-o-u-i-s-g** [reply]
Usually an  icon is just a small drawing  supplied by the framework, like an arrow , a bin,  a reload sign etc .. right?  If that´s the case then how would you make  an image gallery, where when you click on the small image  it brings you the large image?   or do you define  an icon as any small image that can even be created at runtime?

**Reply by freakboy3742** [reply]
An icon doesn't necessarily have to be provided by the OS - they can be user provided. The key detail is that they're small, low detail (i.e., usually only 1 color, or a very restricted palette) and have a 1:1 aspect ratio. The examples you've provided (bin, reload, etc) would all be icons.

An image gallery is an interesting use case - they're definitely *not* icons, so I agree there's a need for something there. Adding an on_click handler to Image would definitely be the simplest approach. My concern is that by making that option available, we'll end up with a flood of GUIs using Image instead of Button because someone wants to "customise" their UI - something that is fundamentally opposed to the design philosophy of Toga.

However, the only alternative I can think of would be an "ImageGallery" widget - something that is a lot more complex as a proposition, and isn't as obvious that it has a "default" implementation. It's also something that *could* be built from other primitives once a clickable Image exists.

**Reply by l-o-u-i-s-g** [reply]
I do appreciate the simplicity of Toga. I agree that an ImageGallery would over complicate things and it may not apply to all use case.  For example 10 years ago I built an app on the android sdk  which allows you to create custom lists at run time.  A list item is basically a label  with a form action and a  widget on each sides. One of these available widget is an Image.  So for example you can make a list medicinal plants.  The label is the name of the plant.  when you click on the label you get the description/benefits of the plant.  the image widget is a small image of the plant. when you click on that image you get  the large image. 

I also understand your point of people that may start using Images where Icons should be used  but I think that issue can be solved with proper documentation.  

and how do we approach  long presses  on mobile and right click on desktop?  LOL

**Reply by freakboy3742** [reply]
The "long press" vs "right click" is one we *do* have an answer for - you capture the UX use case, not the specific physical action. This is one of the reasons that the handler for button is "on_press", not "on_click" - "click" implicitly suggests that a mouse will be used to perform the action. 

In this specific case, something like "on_alternate_press" or "on_context_press" would make sense - something that captures the functional purpose of the action, rather than the platform specifics of the cation. This would then be implemented as a right click on windows, a Control-click on macOS, and a long press on mobile platforms - three different physical manifestations of the underlying UI concept.

**Reply by l-o-u-i-s-g** [reply]
and for linux which  can be desktop or mobile?  There will need to be a way to detect the platform.

**Reply by freakboy3742** [reply]
... you mean, like the tools that Toga already has (`toga.platform.current_platform`)? Toga doesn't consider Android to be a "Linux"; and even if it did, toga_android and toga_gtk are separate backends, so they can have different implementations.

**Reply by l-o-u-i-s-g** [reply]
I actually mean real Linux phones , like the pine phone and librem   or actually any Os  with a touch screen I guess. like windows in tablet mode..

**Reply by freakboy3742** [reply]
I don't know anything about the specifics of PinePhone or Librem devices to be able to comment. However, I can't see how they could be doing anything that would be incompatible with what I've already said. Either (a) there's a need for a separate toga_pine (or whatever) backend, or (b) the Gtk backend (which is the current default backend for Linux) needs to be modified to trigger `on_context_press` on a long touch as well as a right click. While it might be nice to be able to clearly identify Pine/Librem phones from desktop Linux, neither of these approaches would *require* Toga to be able to differentiate a desktop platform from a Pine/Librem phone platform. 

If you're motivated to support PinePhone or Librem devices, then by all means contribute patches to that effect; but I can't say they're an especially high priority on my own personal roadmap.

**Comment by johnzhou721** [comment]
The iOS Photos app presses an image to show a photo.  I think that'd be a very possible usecase, although we need to document that it's *NOT* for doing things like what other people are asking for in this thread as a generic button — but it is a possible usecase.

**Reply by HalfWhitt** [reply]
It's true, image galleries and the like are probably the best argument for true clickable images; see https://github.com/beeware/toga/discussions/1343#discussioncomment-4877254
